gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -262,86 +260,72 @@ impl GatewayLightning for ClnRpcService {
-            };
-
-            // TODO: Implement non-blocking access to the outcomes map sowe can multithread complete_htlcs
-            if let Some(outcome) = self.interceptor.outcomes.lock().await.remove(&hash) {
-                // Translate action request into a cln rpc response for `htlc_accepted` event
-                let htlca_res = match action {
-                    Some(Action::Settle(Settle { preimage })) => {
-                        if let Ok(pk) = XOnlyPublicKey::from_slice(&preimage) {
-                            serde_json::json!({ "result": "resolve", "payment_key": pk.to_string() })
-                        } else {
-                            temp_node_failure()
-                        }
-                    }
-                    Some(Action::Cancel(Cancel { reason: _ })) => {
-                        // TODO: Translate the reason into a BOLT 4 failure message
-                        // See: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages
-                        temp_node_failure()
-                    }
-                    None => {
-                        error!("No action specified for intercepted htlc id: {:?}", hash);
-                        results.push(Err(Status::internal(
-                            "No action specified on this intercepted htlc",
-                        )));
-                        continue;
-                    }
-                };
+        } = request.into_inner();
 
-                // Send translated response to the HTLC interceptor for submission to the cln rpc
-                match outcome.send(htlca_res) {
-                    Ok(_) => {
-                        results.push(Ok(CompleteHtlcsResponse {}));
-                    }
-                    Err(e) => {
-                        error!(
-                            "Failed to send htlc_accepted response to interceptor: {:?}",
-                            e
-                        );
-                        results.push(Err(Status::internal(
-                            "Failed to send htlc_accepted outcome to interceptor",
-                        )));
-                    }
-                };
-                continue;
+        let hash = match sha256::Hash::from_slice(&intercepted_htlc_id) {
+            Ok(hash) => hash,
+            Err(e) => {
+                error!("Invalid intercepted_htlc_id: {:?}", e);
+                return Err(Status::invalid_argument(e.to_string()));
             }
+        };
+
+        // TODO: Implement non-blocking access to the outcomes map sowe can multithread complete_htlcs
+        if let Some(outcome) = self.interceptor.outcomes.lock().await.remove(&hash) {
+            // Translate action request into a cln rpc response for `htlc_accepted` event
+            let htlca_res = match action {
+                Some(Action::Settle(Settle { preimage })) => {
+                    if let Ok(pk) = XOnlyPublicKey::from_slice(&preimage) {
```

Comment:
We actually don't care if it's a valid `XOnlyPublicKey` in the extension afaik, that's the problem of the gateway. We could make it a `[u8; 32]` in a follow-up.

Reply:
This type was originally a `[u8; 32]` Preimage in our definitions. Sending it as Protobuf type over grpc loses the sizing information, but we only need to coerce it into a string to send to cln-pay.

opted to go with `String::from_utf8(preimage)`

Reply:
Why a string though? CLN guarantees this to be 32 hex encoded bytes. My opposition to xonly pub key was that CLN doesn't guarantee its a well-formed pub key and that this is just a fedimint convention and should probably be checked in gatewayd.

Reply:
I mean, technically yes, but only because JSON is shit and doesn't have a native binary data type :laughing: it's a 32bytes hex-encoded array

Reply:
@elsirion, there's no blocking change required here, right? Can we close pr?

Reply:
Just had another look and I think I confused the flow direction of data here actually, so sorry if I caused any confusion.

While I still think not parsing into `XOnlyPublicKey` is semantically more correct, the previous version at least lead to a valid encoding (by accident?) because [`XOnlyPublicKey`'s `Display` impl serializes to hex](https://docs.rs/secp256k1/latest/src/secp256k1/key.rs.html#1127-1129). `String::from_utf8(preimage)` does not achieve this afaik (unless we are transmitting hex encoded data as raw bytes, which would be even more broken).

What we want is taking the `&[u8]` from protobuf and converting it into a hex encoded `[u8; 32]`. So the right move is probably adding an assert that `preimage` is 32bytes long and then using the hex impl form `bitcoin_hashes` (as @mxz42 did all over fedimint recently, e.g. #1472) to generate the string that can be given to CLN.

Reply:
🔥 . Would have been a bad bug. Good catch on the utf8 stuff